cumulusci/core/runtime.py

```diff
@@ -1,23 +1,29 @@
 from abc import abstractmethod
 import sys
 
+from contextvars import ContextVar
+
 from cumulusci.core.config import UniversalConfig, BaseProjectConfig
 from cumulusci.core.exceptions import NotInProject, ProjectConfigNotFound
 from cumulusci.core.keychain import BaseProjectKeychain
 from cumulusci.core.flowrunner import FlowCallback, FlowCoordinator
 
 
+DEBUG_MODE = ContextVar("debug_mode")
+
+
 # pylint: disable=assignment-from-none
 class BaseCumulusCI(object):
     universal_config_class = UniversalConfig
     project_config_class = BaseProjectConfig
     keychain_class = BaseProjectKeychain
     callback_class = FlowCallback
 
-    def __init__(self, *args, load_keychain=True, **kwargs):
+    def __init__(self, *args, load_keychain=True, debug_mode=False, **kwargs):
         self.universal_config = None
         self.project_config = None
         self.keychain = None
+        DEBUG_MODE.set(debug_mode)
```

Comment:
I really don't like using a global variable for this (even one that is context-safe).

When we added the runtime object, the thinking was that we would put any contextual information needed by tasks on the runtime object, and then add a way for tasks to access their runtime. I think we should handle this in that way.

Reply:
I did implement it that way at first but changed my mind.

There are a few features that naturally lend themselves to context vars or globals.

Logging, for example. Python logging is global-variable based because it is unreasonable to thread the log variable through to every corner of the code.  --debug is highly tied to and analogous with logging.

If it were only tasks that need access to it, I'd do it that way, but any algorithm in the whole system might need access, which means it becomes a part of the type signature of every constructor and function in all of CCI. I think that's a bigger problem than using a ContextVar.

Note also that this option is unlikely to be used in multi-RUNTIME contexts like servers.

Reply:
BTW, today I wanted access to the --debug option in `SfdxOrgConfig.sfdx_info`.

I could imagine a property on classes like that self.debug which is opaque about where the data is really coming from. Then we could change it later if needed.

Passing the RUNTIME to tasks is a two-edged sword and might cause even bigger long-term problems. If custom tasks start to depend on having access to basically everything, refactoring could be tricky later.

Reply:
Okay... I concede that there are some things where it's handy to be able to assume they're present as part of a global context and not have to pass them around explicitly.

But, part of what makes me uneasy here is that we already have a few things which are global, and there's no consistency in how we're handling them. RUNTIME is just a plain module global. The CURRENT_TASK stack in cumulusci.core.tasks is a threading.local. Logging is an interesting one (because we should be able to just use global loggers from the logging module and rely on it to route log entries to the right handlers based on the logger name -- but we're actually passing it around in many places that we probably don't need to). Now you're proposing to add a ContextVar. So it seems like maybe it's the right time to try to make things more consistent (possibly consolidate into a single context object, but I do share that concern about things starting to implicitly depend on having access to everything).

Globals are also a bit of a pain to test, since you have to patch them in the correct module where they are used, not where they are defined.

I'd like to hear more about your anticipated use cases for this. If it's just about logging debug information, that's what loglevels are for. So I suspect it's more than that?

Reply:
In many of my use-cases I would get rid of task-specific flags like --keep-working-directory (or whatever its called) and just use a single global debug flag. In particular for data-loading I'd like to provide the user with copious intermediate data (e.g. the raw CSVs imported and exported, the mapping file used, including runtime annotations, etc.)

I suspect that keeping around lots of working files and logs could also be helpful for debugging metadata deploy issues.

ContextVars are just the 21st century way to do Thread-Local Variables, so yeah, now that we've got the module, CURRENT_TASK should definitely be updated (or replaced).

IMO, RUNTIME should probably not be global at all, which is why I didn't just add this flag to it. Assuming that there is only one RUNTIME could cause all sorts of scalability problems in future versions of Pipeline API.

Thankfully, RUNTIME seems to be used only in the one file, so it isn't "really" used as a Global yet and we could get rid of it before someone uses it that way.

IMO, ContextVars are actually only like 10% as bad as "true" globals. The one thing you do not want is that when you refactor your app to be multi-threaded or multi-runtimed that you'll get leaking between the threads or runtimes and ContextVars *should* do a good job of preventing that.

Reply:
Thanks, that's helpful for understanding the use case.

And I agree that (separately) CURRENT_TASK should be switched to a contextvar and RUNTIME should ideally not be a global.

What I'd like to see if we're going to use a contextvar for the debug mode is:
- Move the related bits (DebugMode class, DEBUG_MODE contextvar) into their own module, cumulusci.core.debug_mode
- Provide functions in that module for getting and setting the debug mode. That way the use of a contextvar is an implementation detail, not the actual API. And the current debug mode can always be patched in the same module for tests, rather than needing to remember to patch it where it's imported.
- Set the debug mode directly from the `main` function rather than tying it to BaseCumulusCI's initializer.
- an approved 3pp record for contextvars (it's not enough that it's been approved for some other use; we need an approval for our use)